nuxeo-core/nuxeo-core-storage-dbs/src/main/java/org/nuxeo/ecm/core/storage/dbs/DBSTransactionState.java

```diff
@@ -296,6 +298,67 @@ public boolean hasChild(String parentId, String name) {
         }
         // fetch from repository
         List<State> states = repository.queryKeyValue(KEY_PARENT_ID, parentId, seen);
+        return getDbsDocumentStates(docStates, states);
+    }
+
+    public List<DBSDocumentState> getRegularChildrenStates(String parentId) {
+        List<DBSDocumentState> docStates = new LinkedList<>();
+        Set<String> seen = new HashSet<>();
+        Set<String> specialChildrenTypeNames = getSpecialChildrenTypesNamesSet();
+        // check transient state
+        for (DBSDocumentState docState : transientStates.values()) {
+            if (!parentId.equals(docState.getParentId())) {
+                continue;
+            }
+            if (specialChildrenTypeNames.contains(docState.getPrimaryType())) {
+                continue;
+            }
+            docStates.add(docState);
+            seen.add(docState.getId());
+        }
+        // fetch from repository
+        Map<DBSQueryOperator, List<String>> filter = getSpecialChildrenFilter(DBSQueryOperator.NOT_IN, specialChildrenTypeNames);
+        List<State> states = repository.queryKeyValue(KEY_PARENT_ID, parentId, KEY_PRIMARY_TYPE, filter, seen);
+        return getDbsDocumentStates(docStates, states);
+    }
+
+    public List<DBSDocumentState> getSpecialChildrenStates(String parentId) {
+        List<DBSDocumentState> docStates = new LinkedList<>();
+        Set<String> seen = new HashSet<>();
+        Set<String> specialChildrenTypeNames = getSpecialChildrenTypesNamesSet();
+        // check transient state
+        for (DBSDocumentState docState : transientStates.values()) {
+            if (!parentId.equals(docState.getParentId())) {
+                continue;
+            }
+            if (!specialChildrenTypeNames.contains(docState.getPrimaryType())) {
+                continue;
+            }
+            docStates.add(docState);
+            seen.add(docState.getId());
+        }
+        // fetch from repository
+        Map<DBSQueryOperator, List<String>> filter = getSpecialChildrenFilter(DBSQueryOperator.IN, specialChildrenTypeNames);
+        List<State> states = repository.queryKeyValue(KEY_PARENT_ID, parentId, KEY_PRIMARY_TYPE, filter, seen);
+        return getDbsDocumentStates(docStates, states);
+    }
+
+    protected Map<DBSQueryOperator, List<String>> getSpecialChildrenFilter(DBSQueryOperator operator,
+            Set<String> excludedFromCopyDocumentTypes) {
+        List<String> excludedFromCopyDocumentTypesAsList = //
+                excludedFromCopyDocumentTypes.stream().collect(Collectors.toList());
+        return Collections.singletonMap(operator, excludedFromCopyDocumentTypesAsList);
+    }
+
+    protected Set<String> getSpecialChildrenTypesNamesSet() {
+        return Framework.getService(SchemaManager.class)
+                        .getSpecialDocumentTypes()
+                        .stream()
+                        .map(s -> s.getName())
```

Comment:
`.map(DocumentType::getName)`

Reply:
But if all we need is the type names, then our new API `getSpecialDocumentTypes` should exactly that (as a `Set`), not a list of `DocumentType`.